url3426a97ee4_rtreffer_minidns_tgz-pJ8-de_measite_minidns_minidnsrepl_MiniDnsReplJ8/src/de/measite/minidns/recursive/RecursiveDNSClient.java

```diff
@@ -0,0 +1,597 @@
+
+        switch (ipVersionSetting) {
+        case v4only:
+            for (A a : getCachedIPv4NameserverAddressesFor(parent)) {
+                if (primaryTarget == null) {
+                    primaryTarget = a.getInetAddress();
+                    continue;
+                }
+                secondaryTarget = a.getInetAddress();
+                break;
+            }
+            break;
+        case v6only:
+            for (AAAA aaaa : getCachedIPv6NameserverAddressesFor(parent)) {
+                if (primaryTarget == null) {
+                    primaryTarget = aaaa.getInetAddress();
+                    continue;
+                }
+                secondaryTarget = aaaa.getInetAddress();
+                break;
+            }
+            break;
+        case v4v6:
+            InetAddress[] v4v6targets = getTargets(getCachedIPv4NameserverAddressesFor(parent), getCachedIPv6NameserverAddressesFor(parent));
+            primaryTarget = v4v6targets[0];
+            secondaryTarget = v4v6targets[1];
+            break;
+        case v6v4:
+            InetAddress[] v6v4targets = getTargets(getCachedIPv6NameserverAddressesFor(parent), getCachedIPv4NameserverAddressesFor(parent));
+            primaryTarget = v6v4targets[0];
+            secondaryTarget = v6v4targets[1];
+            break;
+        default:
+            throw new AssertionError();
+        }
+
+        DNSName authoritativeZone = parent;
+        if (primaryTarget == null) {
+            authoritativeZone = DNSName.ROOT;
+            switch (ipVersionSetting) {
+            case v4only:
+                primaryTarget = getRandomIpv4RootServer();
+                break;
+            case v6only:
+                primaryTarget = getRandomIpv6RootServer();
+                break;
+            case v4v6:
+                primaryTarget = getRandomIpv4RootServer();
+                secondaryTarget = getRandomIpv6RootServer();
+                break;
+            case v6v4:
+                primaryTarget = getRandomIpv6RootServer();
+                secondaryTarget = getRandomIpv4RootServer();
+                break;
+            }
+        }
+
+        List<IOException> ioExceptions = new LinkedList<>();
+
+        try {
+            return queryRecursive(recursionState, q, primaryTarget, authoritativeZone);
+        } catch (IOException ioException) {
+            abortIfFatal(ioException);
+            ioExceptions.add(ioException);
+        }
+
+        if (secondaryTarget != null) {
+            try {
+                return queryRecursive(recursionState, q, secondaryTarget, authoritativeZone);
+            } catch (IOException ioException) {
+                ioExceptions.add(ioException);
+            }
+        }
+
+        MultipleIoException.throwIfRequired(ioExceptions);
+        return null;
+    }
+
+    private DNSMessage queryRecursive(RecursionState recursionState, DNSMessage q, InetAddress address, DNSName authoritativeZone) throws IOException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 257-282, lines 285-316 into separate methods.